pkg/p2p/streamtest/streamtest.go

```diff
@@ -130,12 +134,16 @@ func (r *Record) setErr(err error) {
 }
 
 type stream struct {
-	in  io.WriteCloser
-	out io.ReadCloser
+	in     io.WriteCloser
+	out    io.ReadCloser
+	cin    chan struct{}
+	cout   chan struct{}
+	closed bool
+	mtx    sync.Mutex // guards close
```

Comment:
I think that this can be replaced with `closeOnce sync.Once`, and to use it in Close method, for the whole function body.